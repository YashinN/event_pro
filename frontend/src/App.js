import { BrowserRouter, Routes, Route, Navigate } from "react-router-dom";
import { useState, useEffect } from "react";

// comonents.
import Navbar from "./components/Navbar";
import LoginPage from "./pages/LoginPage";
import HomePage from "./pages/HomePage";
import CreateEventPage from "./pages/CreateEventPage";
import EditEventPage from "./pages/EditEventPage";

function App() {
  // checks if local storage for user.
  const setUser = () => {
    const user = JSON.parse(localStorage.getItem("user"));
    // checks if user exists.
    if (user) {
      // gets user
      return user;
    }
    // return no user
    return null;
  };

  // state to define user.
  const [isUser, setIsUser] = useState(setUser());
  // state to handle logged in or not
  const [loggedIn, setLoggedIn] = useState(null);
  // stores all events.
  const [events, setEvents] = useState([]);
  // stores users saved events.
  const [myEvents, setMyEvents] = useState([]);
  // loading state for api calls.
  const [loading, setLoading] = useState(true);

  // request to get all events from server.
  const getEvents = async () => {
    setLoading(true);
    try {
      // req to fetch data
      const response = await fetch(process.env.REACT_APP_EVENTS);
      // stores events as json
      const data = await response.json();

      // if response ok sets component data
      if (response.ok) {
        setLoading(null);
        setEvents(data);
      }
    } catch (error) {
      console.log(error);
    }
  };

  // request to get user saved events from server.
  const getMyEvents = async () => {
    try {
      // request to server with user id as identifier.
      const response = await fetch(
        process.env.REACT_APP_MYEVENTS + isUser.email,
        {
          method: "GET",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${isUser.token}`,
          },
        }
      );
      // stores user saved events as json
      const data = await response.json();
      if (response.ok) {
        // sets component data for user events.
        setMyEvents(data);
      }
    } catch (error) {
      console.log(error);
    }
  };

  // req to verify google and github login
  const getGoogleUser = async () => {
    // fetch req to server/
    const response = await fetch(process.env.REACT_APP_SUCCESS, {
      method: "GET",
      credentials: "include",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
        "Access-Control-Allow-Credentials": true,
      },
    });

    // stores user details.
    const data = await response.json();
    if (response.ok) {
      // gets user email address
      const user = data.user["emails"];
      const email = user[0].value;
      const isAdmin = data.isAdmin;
      const token = data.token;
      // stores user data in local storage.
      localStorage.setItem(
        "user",
        JSON.stringify({ email: email, isAdmin, token })
      );
      // sets users
      setIsUser(setUser());
    }
  };

  // use effect run everytime a user logs in/out or sign up
  useEffect(() => {
    // gets user from local storage.
    const user = JSON.parse(localStorage.getItem("user"));
    if (user) {
      // if user exists sets isUser state.
      setIsUser(user);
    } else {
      setIsUser(null);
    }
  }, [loggedIn]);

  // use effect checks if google or github has been signed in
  // gets all events
  useEffect(() => {
    if (!isUser) {
      getGoogleUser();
    }

    getEvents();
  }, []);

  // useEffect gets user saved events when user logs in or when events state changes.
  useEffect(() => {
    if (isUser) {
      getMyEvents();
    }
  }, [isUser, events]);

  return (
    <div className="App">
      <BrowserRouter>
        <Navbar
          user={isUser}
          setLoggedIn={setLoggedIn}
          events={events}
          myEvents={myEvents}
          setMyEvents={setMyEvents}
        />
        <div className="pages">
          <Routes>
            <Route
              path="/"
              element={
                <HomePage
                  events={events}
                  isUser={isUser}
                  setIsUser={setIsUser}
                  loading={loading}
                  setMyEvents={setMyEvents}
                />
              }
            />
            <Route
              path="/createEvent"
              element={
                !isUser ? (
                  <Navigate to="/" />
                ) : isUser.isAdmin ? (
                  <CreateEventPage
                    setEvents={setEvents}
                    events={events}
                    user={isUser}
                  />
                ) : (
                  <Navigate to="/" />
                )
              }
            />
            <Route
              path="/editEvent"
              element={
                !isUser ? (
                  <Navigate to="/" />
                ) : isUser.isAdmin ? (
                  <EditEventPage
                    events={events}
                    setEvents={setEvents}
                    loading={loading}
                    user={isUser}
                  />
                ) : (
                  <Navigate to="/" />
                )
              }
            />
            <Route
              path="/login"
              element={
                isUser ? (
                  <Navigate to="/" />
                ) : (
                  <LoginPage setLoggedIn={setLoggedIn} />
                )
              }
            />
            <Route path="*" element={<HomePage events={events} />} />
          </Routes>
        </div>
      </BrowserRouter>
    </div>
  );
}

export default App;
