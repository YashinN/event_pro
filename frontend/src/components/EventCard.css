.event-card-img {
  width: 100%;
  height: 200px;
  object-fit: cover;
}

.event-card {
  border-radius: 1rem;
  overflow: hidden;
  position: relative;
}

.color-1 {
  background-color: #15b5b0;
}

.color-2 {
  background-color: #f9bdc0;
}

.color-3 {
  background-color: #fbe698;
}

.color-4 {
  background-color: #ece3f0;
}

.mark-btn {
  position: absolute;
  bottom: 20px;
  right: 10px;
}
