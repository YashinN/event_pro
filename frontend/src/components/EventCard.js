import { useState } from "react";
// modal component
import MessageModal from "./MessageModal";
// component css
import "./EventCard.css";

const EventCard = (props) => {
  // toggles state to show message modal
  const [showModal, setShowModal] = useState(false);
  // sets loading state when making api calls.
  const [loading, setLoading] = useState(null);

  // request to server to save and event to My events or user events.
  const saveRequest = async () => {
    // sets loading state while waiting for response.
    setLoading(true);
    // makes request with user email as identifier in req params
    // sends event id to derver in body.
    const response = await fetch(
      process.env.REACT_APP_MYEVENTS + props.isUser.email,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${props.isUser.token}`,
        },
        body: JSON.stringify({ id: props.event._id }),
      }
    );

    // recevies user saved event data from server
    const data = await response.json();

    // if successful response
    if (response.ok) {
      // toggles off loading state.
      setLoading(null);
      // stores recevied user events
      // props.setIsUser(data);
      props.setMyEvents(data.allUserEvents);

      openModal();
    }
    setLoading(null);
  };

  // onclick request to save user event.
  const handleSave = () => {
    saveRequest();
  };

  // opens message modal on successful save
  const openModal = () => {
    setShowModal(true);
  };

  // closes message modal on click.
  const closeModal = () => {
    setShowModal(false);
  };

  return (
    <div className="col-12 col-md-6 col-lg-6 col-xl-4">
      <div className={`container event-card px-0 ${props.colour}`}>
        {props.isUser && (
          <button
            type="button"
            className="btn btn-outline-primary mark-btn btn"
            onClick={handleSave}
          >
            {loading ? (
              <div className="loading-container">
                <div className="spinner-border" role="status">
                  <span className="visually-hidden">Loading...</span>
                </div>
              </div>
            ) : (
              "Save"
            )}
          </button>
        )}

        <img src={props.event.image} alt="" className="event-card-img" />

        <div className=" p-3 ">
          <h5 className="mb-2">{props.event.description}</h5>
          <h5 className="mb-2">Date:{props.event.date}</h5>
          <h5 className="mb-2">Venue:{props.event.venue}</h5>
          <h5 className="mb-2">
            Time:{props.event.startTime} - {props.event.endTime}
          </h5>
        </div>
      </div>
      <MessageModal
        showModal={showModal}
        closeModal={closeModal}
        message={"Event has been saved!"}
      />
    </div>
  );
};

export default EventCard;
