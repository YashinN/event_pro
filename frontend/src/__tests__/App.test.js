import renderer from "react-test-renderer";
import App from "../App";
import EventCard from "../components/EventCard";

// test if app renders correctly
test("Display renders correclty", () => {
  const tree = renderer.create(<App></App>).toJSON;
  expect(tree).toMatchSnapshot();
});
