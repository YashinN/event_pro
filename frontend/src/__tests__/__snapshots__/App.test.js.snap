// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display renders correclty 1`] = `[Function]`;
